### PR TITLE
Kiln fix gossip policy

### PR DESCRIPTION
**Pull request by MarekM25**

### CODE REVIEW

File: src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs [hunk 1]
```diff
@@ -24,23 +24,50 @@ namespace Nethermind.Merge.Plugin
...
-            _preMergeGossipPolicy = apiGossipPolicy ?? throw new ArgumentNullException(nameof(apiGossipPolicy));
-            _poSSwitcher = poSSwitcher ?? throw new ArgumentNullException(nameof(poSSwitcher));
+            ShouldGossipBlocks = apiGossipPolicy?.ShouldGossipBlocks ?? throw new ArgumentNullException(nameof(apiGossipPolicy));
             _blockFinalizationManager = blockFinalizationManager;
+            
+            _blockFinalizationManager.BlocksFinalized += OnBlockFinalized;
         }
 
-        public bool ShouldGossipBlocks => !_poSSwitcher.HasEverReachedTerminalBlock() && _preMergeGossipPolicy.ShouldGossipBlocks;
+        // According to spec (https://github.com/ethereum/EIPs/blob/d896145678bd65d3eafd8749690c1b5228875c39/EIPS/eip-3675.md#network)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs:39]: Not finalized but terminal

The networking stack SHOULD NOT send the following messages if they advertise the descendant of any terminal PoW block:

[NewBlockHashes (0x01)](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01)
[NewBlock (0x07)](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblock-0x07)

File: src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs [hunk 2]
```diff
@@ -24,23 +24,50 @@ namespace Nethermind.Merge.Plugin
...
 
-        public bool ShouldDisconnectGossipingNodes => _blockFinalizationManager.LastFinalizedHash != Keccak.Zero && _preMergeGossipPolicy.ShouldDisconnectGossipingNodes;
+        // According to spec (https://github.com/ethereum/EIPs/blob/d896145678bd65d3eafd8749690c1b5228875c39/EIPS/eip-3675.md#network)
+        // we MUST stop gossiping peers after receiving next finalized block to FIRST_FINALIZED_BLOCK, so one block after we started discarding NewBlock/NewBlockHash messages
+        public bool ShouldGossipBlocks { get; private set; }
 
+        // According to spec (https://github.com/ethereum/EIPs/blob/d896145678bd65d3eafd8749690c1b5228875c39/EIPS/eip-3675.md#network)
+        // we SHOULD start disconnecting gossiping peers after receiving next finalized block to FIRST_FINALIZED_BLOCK, so one block after we started discarding NewBlock/NewBlockHash messages
+        // for now hardcoded false - in spec it is SHOULD, not MUST, so we don't want to disconnect any peers during transition. Later just change to ShouldGossipBlocks
+        public bool ShouldDisconnectGossipingNodes => false;    //=> ShouldGossipBlocks;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs:50]: we can use _mergeConfig.FinalTotalDifficulty != null here? if we set this value it means that we're post transition

File: src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs [hunk 3]
```diff
@@ -24,23 +24,50 @@ namespace Nethermind.Merge.Plugin
...
+        public bool ShouldDiscardBlocks { get; private set; }
 
-        public bool ShouldDisconnectGossipingNodes => _blockFinalizationManager.LastFinalizedHash != Keccak.Zero && _preMergeGossipPolicy.ShouldDisconnectGossipingNodes;
+        // According to spec (https://github.com/ethereum/EIPs/blob/d896145678bd65d3eafd8749690c1b5228875c39/EIPS/eip-3675.md#network)
+        // we MUST stop gossiping peers after receiving next finalized block to FIRST_FINALIZED_BLOCK, so one block after we started discarding NewBlock/NewBlockHash messages
+        public bool ShouldGossipBlocks { get; private set; }
 
+        // According to spec (https://github.com/ethereum/EIPs/blob/d896145678bd65d3eafd8749690c1b5228875c39/EIPS/eip-3675.md#network)
+        // we SHOULD start disconnecting gossiping peers after receiving next finalized block to FIRST_FINALIZED_BLOCK, so one block after we started discarding NewBlock/NewBlockHash messages
+        // for now hardcoded false - in spec it is SHOULD, not MUST, so we don't want to disconnect any peers during transition. Later just change to ShouldGossipBlocks
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs:49]: we need tests and let's talk :)

File: src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs (hunk 2 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs]: done

File: src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs [hunk 4]
```diff
@@ -16,31 +16,42 @@
...
 
-        public bool ShouldGossipBlocks => !_poSSwitcher.HasEverReachedTerminalBlock() && _preMergeGossipPolicy.ShouldGossipBlocks;
+        // According to spec (https://github.com/ethereum/EIPs/blob/d896145678bd65d3eafd8749690c1b5228875c39/EIPS/eip-3675.md#network)
+        // We SHOULD NOT advertise the descendant of any terminal PoW block.
+        // In our approach we CAN gossip blocks if we didn't received FIRST_FINALIZED_BLOCK yet...
+        public bool CanGossipBlocks => !_poSSwitcher.TransitionFinished && _preMergeGossipPolicy.CanGossipBlocks;
 
-        public bool ShouldDisconnectGossipingNodes => _blockFinalizationManager.LastFinalizedHash != Keccak.Zero && _preMergeGossipPolicy.ShouldDisconnectGossipingNodes;
+        //  ...and gossipping policy will be decided according to the header difficulty.
+        public bool ShouldGossipBlock(BlockHeader header) => !_poSSwitcher.GetBlockConsensusInfo(header).IsPostMerge;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs:46]: add condition with TransitionFinished or _mergeConfig.FinalTotalDifficulty?

Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Merge.Plugin/MergeGossipPolicy.cs:46]: not needed here. It's used only once, only if CanGossip is true, so TransitionFinished is already checked
https://github.com/NethermindEth/nethermind/blob/c10515bbd76763e443d64e8750094393969b23ba/src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs#L280

**Pull request merged**